### ISSUE TITLE
Support `package-lock.json` file

### ISSUE THREAD

**Issue by piranna** [npmgraph/npmgraph#148]
Allow to provide a `package-lock.json` file instead of a `package.json` file to draw the graph. This can be useful when using private packages not published on npmjs registry, or using other registries like Github Packages Registry, or when using git dependencies.

Alternatively, or complementary to that, allow to provide tokens to access to that other registries or git repositories.

**Comment by broofa** [comment]
Closing.  Encoding  100KB+ lock files into the URL hash - which is what we currently do with package.json content - is going to be problematic.

Also, I just don't see a great deal of incremental benefit to this.  I know it'll show the exact installed packages rather than the most recent semver-matching ones, but typically that difference isn't very interesting in the npmgraph sense.  (minor-version diffferences don't generally affect the npmgraph diagrams all that much.)

**Comment by piranna** [comment]
Using the URL hash is not so good, since URLs length is at most 2KB, so i'm surprised It works at all with somewhat big package.json files.

Regarding minor versions, it's true that It would not have too much differences for published packages, but It will have them for git dependencies or when versioning allows to upgrade Major versions. If it's not possible to print someway the real graph, i find It as a bug. So, how can we fix It?

**Comment by broofa** [comment]
> So, how can we fix It?

Well... one option would be to accept URLs to lock files like we do for package.json files, [like this](https://npmgraph.js.org/?q=https%3A%2F%2Fgithub.com%2Fnpmgraph%2Fnpmgraph%2Fblob%2Fmain%2Fpackage-lock.json).  (Only shows one module currently because npmgraph doesn't grok package-lock files [yet])

That would circumvent the issue of where to store the lock file content.

**Comment by piranna** [comment]
But that would not work for big lock files... I think adding an input element to be able to load the package or lock files could work, in addition to give the actual content on the URL.

**Comment by broofa** [comment]
> an input element to be able to load the package

Load _where_ though?  The issue here isn't the UI.  It's that (so far) there's no backend server.  That's a design choice I made (for better or worse) many years ago.  Changing that would likely require adding a way for users to authenticate, as well as worrying about pricing and plan limits, etc.

Alternatively, there could be a self-hosted version of npmgraph that supports connecting to a local DB.   But that's not exactly simple either.

tl;dr: This is an architecture limitation, not a UI limitation.

**Comment by broofa** [comment]
> Load in memory in the browser, there's no need to upload the file anywhere

Such diagrams wouldn't be shareable, and the ability to share (i.e. _communicate_ about dependency concerns) is one of the main selling points of npmgraph.

**Comment by broofa** [comment]
> strip It down to the minimum required info

This is an interesting idea, but still has some challenges.  For example, if you take [this project's package-lock.json file](https://github.com/npmgraph/npmgraph/blob/main/package-lock.json) and distill it down to just the `dependencies` mentioned therein, the names alone (no semver specifiers) are still ~12KB of data (3.8KB gzip'ed).  That's still kind of a lot.

One option would be to only record the top-level dependencies and assume that everything below that the transitive dependencies haven't been changed.  That would actually be pretty compact (a few hundred bytes), but wouldn't capture direct edits to package-lock files (e.g. [like this @dependabot commit](https://github.com/npmgraph/npmgraph/pull/320/files).)

**Comment by piranna** [comment]
We only need the `resolved` field, that's the actually resolved and fetched packages, we have there all the needed info details. Also, we could be able to optionally ignore the development packages, because they are not being installed when the packages is installed as a dependency, reducing the needed ones.

**Comment by broofa** [comment]
I think we're saying basically the same thing: be as smart as possible about using the actually-installed version of each package in the graph as captured in the package-lock file.  Unfortunately even an optimized form of that (only store name + version, omit dev dependencies, etc.) will stil be uncomfortably large for most projects to put in the url hash (10KB+).

FWIW, I briefly toyed with the idea of only recording the version for packages that differ from what a fresh install would use, but that doesn't work either.  Even if the installed version and the default version are the same today, that may not be true tomorrow.  (e.g. a new version gets published.)   You really do need to record every version of every module.  (Not too surprising since preventing this kind of ambiguity is the whole point of lock files.)
